exporter/internal/queue/batcher.go

```diff
@@ -0,0 +1,145 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package queue // import "go.opentelemetry.io/collector/exporter/internal/queue"
+
+import (
+	"context"
+	"math"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/exporter/exporterbatcher"
+	"go.opentelemetry.io/collector/exporter/internal"
+)
+
+type batch struct {
+	ctx     context.Context
+	req     internal.Request
+	idxList []uint64
+}
+
+type Batcher struct {
+	batchCfg exporterbatcher.Config
+
+	queue      Queue[internal.Request]
+	maxWorkers int
+
+	exportFunc func(context.Context, internal.Request) error
+
+	currentBatch *batch // the batch that is being built
+	timer        *time.Timer
+	shutdownCh   chan bool
+
+	stopWG sync.WaitGroup
+}
+
+func NewBatcher(batchCfg exporterbatcher.Config, queue Queue[internal.Request],
+	maxWorkers int, exportFunc func(context.Context, internal.Request) error) *Batcher {
+	return &Batcher{
+		batchCfg:   batchCfg,
+		queue:      queue,
+		maxWorkers: maxWorkers,
+		exportFunc: exportFunc,
+		stopWG:     sync.WaitGroup{},
+		shutdownCh: make(chan bool, 1),
+	}
+}
+
+// If preconditions pass, flush() take an item from the head of batch list and exports it.
+func (qb *Batcher) flush(batchToFlush batch) {
+	err := qb.exportFunc(batchToFlush.ctx, batchToFlush.req)
+	for _, idx := range batchToFlush.idxList {
+		qb.queue.OnProcessingFinished(idx, err)
+	}
+}
+
+// flushAsync() starts a goroutine that calls flushIfNecessary(). It blocks until a worker is available.
+func (qb *Batcher) flushAsync(batchToFlush batch) {
+	// maxWorker = 0 means we don't limit the number of flushers.
+	if qb.maxWorkers == 0 {
+		qb.stopWG.Add(1)
+		go func() {
+			qb.flush(batchToFlush)
+			qb.stopWG.Done()
+		}()
+		return
+	}
+	panic("not implemented")
+}
+
+// Start ensures that queue and all consumers are started.
+func (qb *Batcher) Start(ctx context.Context, host component.Host) error {
+	if err := qb.queue.Start(ctx, host); err != nil {
+		return err
+	}
```

Comment:
That's a good point. Let me keep them here for now because this is consistent with `queue.consumer`, but I'll leave a note to move this up to queue sender once `queue.consumer` is cleaned up.

Reply:
https://github.com/open-telemetry/opentelemetry-collector/pull/11534

Reply:
PR is merged. We can remove this I guess